eth/stagedsync/stage_headers.go

```diff
@@ -236,10 +255,16 @@ Loop:
 		}
 		// Load headers into the database
 		var inSync bool
-		if inSync, err = cfg.hd.InsertHeaders(headerInserter.NewFeedHeaderFunc(tx, cfg.blockReader), cfg.chainConfig.TerminalTotalDifficulty, logPrefix, logEvery.C, uint64(currentTime.Unix())); err != nil {
+		borPenalties := make([]headerdownload.PenaltyItem, 0, 1)
+		if inSync, err = cfg.hd.InsertHeaders(headerInserter.NewFeedHeaderFunc(tx, cfg.blockReader), &borPenalties, cfg.chainConfig.TerminalTotalDifficulty, logPrefix, logEvery.C, uint64(currentTime.Unix())); err != nil {
```

Comment:
passing borPenalties through the call chain into `InsertHeader` feels too intrusive
another way of doing it would be to add an extra return parameter to `InsertHeader` that returns all unverified link headers
then borPenalties could be computed based on that here